### PR TITLE
Re-land "Make go autobumper configurable so it can work with more tools" 

### PR DESCRIPTION
**Pull request by mpherman2**
...with changes to allow prefixes to have different consistency flags. This allows us to specify that k8s/testimages prefixes are not consistently bumped. This was stopping the autobumper the first time around.

### CODE REVIEW

File: experiment/autobumper/bumper/bumper.go [hunk 1]
```diff
@@ -555,24 +551,43 @@ func HasChanges() (bool, error) {
...
+func getPrefixesString(prefixes []Prefix) string {
+	return strings.Join(getAllPrefixes(prefixes), ", ")
+}
+
+func makeCommitSummary(prefixes []Prefix, versions map[string][]string) string {
+	if len(versions) == 0 {
+		return fmt.Sprintf("Update %s images as necessary", getPrefixesString(prefixes))
 	}
-	return fmt.Sprintf("Update prow to %s, and other images as necessary.", version), nil
+	inconsistentBumps := fmt.Sprintf("")
```
Comment (MushuEE) [inline experiment/autobumper/bumper/bumper.go:562]: Curious what the `fmt.Sprintf(` does on these lines?

File: experiment/autobumper/bumper/bumper_test.go [hunk 2]
```diff
@@ -760,44 +807,104 @@ func TestCDToRootDir(t *testing.T) {
...
+				t.Errorf("expected error but did not get one")
 			}
-			if errMsg != tc.expectedError {
-				t.Fatalf("got error '%v', expected error %s", err, tc.expectedError)
+			if !tc.err && err != nil {
+				t.Errorf("expected no error, but got one: %v", err)
 			}
-			if err != nil {
-				return
+			if !reflect.DeepEqual(tc.expectedVersions1, versions) && !reflect.DeepEqual(tc.expectedVersions2, versions) {
```
Comment (michelle192837) [inline experiment/autobumper/bumper/bumper_test.go:860]: I think cmp.Diff should be easier here, if comparing the maps is an issue.

Comment (mpherman2) [reply experiment/autobumper/bumper/bumper_test.go:860]: DeepEqual seems to be working fine. The only issue is that sometimes the map returned is {a: [x, y]} and other times its {a: [y, x]} and I couldn't find a way around that other than checking both. cmp.Diff does not seem to help in this case. Any ideas on how to make the check easier?

Comment (michelle192837) [reply experiment/autobumper/bumper/bumper_test.go:860]: cmp.Diff has options you can pass to ignore order (I think "cmpopts.SortSlices" to compare slices ignoring order? https://godoc.org/github.com/google/go-cmp/cmp/cmpopts#SortSlices), which should be an easier comparison (especially if these tests ever get more complex).

Comment (mpherman2) [reply experiment/autobumper/bumper/bumper_test.go]: This is perfect! Thank you! :)

**Review by michelle192837 — APPROVED** [review]
Think it looks good!

File: experiment/autobumper/bumper/bumper_test.go [hunk 3]
```diff
@@ -760,44 +810,102 @@ func TestCDToRootDir(t *testing.T) {
...
 			}
-			if errMsg != tc.expectedError {
-				t.Fatalf("got error '%v', expected error %s", err, tc.expectedError)
+			if !tc.err && err != nil {
+				t.Errorf("expected no error, but got one: %v", err)
 			}
-			if err != nil {
-				return
+			if diff := cmp.Diff(tc.expectedVersions, versions, cmpopts.SortSlices(func(x, y string) bool { return strings.Compare(x, y) > 0 })); diff != "" {
+				t.Errorf("versions returned unexpected value:\n%s", diff)
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply experiment/autobumper/bumper/bumper_test.go]: Done

**Review by michelle192837 — APPROVED** [review]

File: experiment/autobumper/bumper/bumper.go [hunk 4]
```diff
@@ -206,9 +234,7 @@ func Run(o *Options) error {
 		return fmt.Errorf("failed to change to root dir: %w", err)
 	}
 
-	images, err := UpdateReferences(
-		o.BumpProwImages, o.BumpBoskosImages, o.BumpTestImages, o.TargetVersion,
-		o.IncludedConfigPaths, o.ExcludedConfigPaths, o.ExtraFiles)
+	images, err := UpdateReferences(o)
```
Comment (alvaroaleman) [inline experiment/autobumper/bumper/bumper.go:237]: This now takes an options struct with 15 parameters, most of which used to be constants, that has no defaulting. Can we in the future please be a bit thoughtful about backwards compatibility when changing exported interfaces?

File: experiment/autobumper/bumper/bumper.go [hunk 5]
```diff
@@ -94,25 +83,54 @@ func (af *fileArrayFlag) Set(value string) error {
...
+	//The name used in the address when creating remote. Format will be git@github.com:{GitLogin}/{RemoteName}.git
+	RemoteName string `yaml:"remoteName"`
+	//List of prefixes that the autobumped is looking for, and other information needed to bump them
+	Prefixes []Prefix `yaml:"prefixes"`
+}
+
+// Prefix is the information needed for each prefix being bumped.
+type Prefix struct {
+	//Name of the tool being bumped
+	Name string `yaml:"name"`
```
Comment (alvaroaleman) [inline experiment/autobumper/bumper/bumper.go:121]: In the future, never use `yaml` tags, use `json` tags, bceause they are unerstood by both the yaml and the json serializer

File: experiment/autobumper/bumper/bumper.go [hunk 6]
```diff
@@ -94,25 +83,54 @@ func (af *fileArrayFlag) Set(value string) error {
...
+	//"The email to use on the git commit. Requires GitName. If not specified, uses values from the user associated with the access token.
+	GitEmail string `yaml:"gitEmail"`
+	//The oncall address where we can get the JSON file that stores the current oncall information.
+	OncallAddress string `yaml:"onCallAddress"`
+	//Whether to skip creating the pull request for this bump.
+	SkipPullRequest bool `yaml:"skipPullRequest"`
+	//The URL where upstream images are located. Must not be empty if Target Version is "upstream" or "upstreamStaging"
+	UpstreamURLBase string `yaml:"upstreamURLBase"`
+	//The config paths to be included in this bump, in which only .yaml files will be considered. By default all files are included.
+	IncludedConfigPaths []string `yaml:"includedConfigPaths"`
```
Comment (alvaroaleman) [inline experiment/autobumper/bumper/bumper.go:105]: in the future, please use snake case for serialized field names, like the rest of the codebase does. Now probably too late, as there is already a config for this

**Pull request merged**